src/rt/sched/cown.h

```diff
@@ -47,46 +47,40 @@ namespace verona::rt
 
   struct ReadRefCount
   {
+  private:
     std::atomic<size_t> count{0};
 
-    void add_read()
+  public:
+    // true means first reader is added, false otherwise
+    bool add_read(int readers = 1)
     {
-      count.fetch_add(2);
+      return count.fetch_add(readers * 2, std::memory_order_release) == 0;
     }
 
-    // true means last reader and writer is waiting, false otherwise
+    bool any_reader()
+    {
+      return count.load(std::memory_order_acquire) != 0;
+    }
+
+    // true means last reader is waiting, false otherwise
     bool release_read()
     {
-      if (count.fetch_sub(2) == 3)
+      if (count.fetch_sub(2, std::memory_order_release) == 2)
       {
         Systematic::yield();
-        assert(count.load() == 1);
-        count.store(0, std::memory_order_relaxed);
         return true;
       }
       return false;
     }
 
-    bool try_write()
+    size_t get_count()
     {
-      if (count.load(std::memory_order_acquire) == 0)
-        return true;
-
-      // Mark a pending write
-      if (count.fetch_add(1) != 0)
-        return false;
-
-      // if in the time between reading and writing the ref count, it
-      // became zero, we can now process the write, so clear the flag
-      // and continue
-      count.fetch_sub(1);
-      Systematic::yield();
-      assert(count.load() == 0);
-      return true;
+      return count.load(std::memory_order_acquire);
     }
   };
```

Comment:
I think this deleted code might be useful for addressing the remaining ABA problem.